### DISCUSSION TITLE
Jupyterlite server extension plugin failed to activate

### DISCUSSION THREAD

**Discussion by ashok777** [jupyterlite/jupyterlite#400] [Q&A] [answered]
1. Followed the directions in the developer guide to set up a Server Extension.
2. Cookiecutter generated the extension artifacts - including a single source artifact - a src\index.ts file. Did not make any changes to index.ts.
3. Linked the development extension and ran jlmp build successfully.
4. ran : jupyter lite build --force
5. ran: jupyter lite serve
    server started successfully; extension activated sucessfully as seen in the console.
6. Added file src\kernel.ts (exact copy of https://github.com/jupyterlite/jupyterlite/tree/main/packages/javascript-kernel/src/kernel.ts).
7. Replaced src/index.ts with the index.ts from https://github.com/jupyterlite/jupyterlite/tree/main/packages/javascript-kernel-extension/src.
    7.1 Changed the plugin id to reflect my extension name ('jupyterlite-lexo-ext:plugin');
    7.2 imported the local kernel thusly: import { JavaScriptKernel } from './kernel';

8. jlpm build ran successfully.
9. rebuilt the site: jupyter lite build --force
10. ran: jupyter lite serve
   server started successfully; but the following messages seen in console:
     **Plugin failed to activate; Error: No provider for: @jupyterlite/kernel:IKernelSpecs.**

The original cookiecutter generated src\index.ts depended on '@jupyterlite/server' and it was resolved successfully. Now, after modification, there is a dependency on '@jupyterlite/kernel', and that is not getting resolved at run time even though the build was successful.

Any help in the resolution of this issue is greatly appreciated.

**Comment by jtpio** [comment, marked as answer]
Thanks @ashok777 for trying it out and reporting here.

It sounds like this piece of configuration might be missing from the `package.json`? 

https://github.com/jupyterlite/echo-kernel/blob/4310974489efbc684247860a6f573e88449ab2e4/package.json#L76-L81

It's not part of the cookiecutter because the cookiecutter doesn't add a new kernel. But maybe we should document it in https://jupyterlite.readthedocs.io/en/latest/extensions.html. There are a couple of explanations about this concept in the JupyterLab documentation (https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#deduplication), which are for the most part applicable here too.

**Reply by ashok777** [reply]
Thanks very much @jtpio. That was exactly what the doctor ordered. I now have a working kernel extension and can focus on the changes I need to make to get the desired functionality.

**Reply by jtpio** [reply]
Nice!
